Use rclcpp::guard_condition

mauropasse opened the pull request:
This PR makes **rclcpp** to not handle directly `rcl_guard_condition_t`, but to use the `rclcpp::GuardCondition` class which takes care of creation/init/delete of it, which makes the code cleaner and less error prone. 

The work being done in https://github.com/ros2/rmw/pull/286 assigns listener callbacks to RMW entities, and originally also assigned a callback to the RMW guard conditions, since many events are generated from the triggering of a guard condition. The listener callback for guard conditions is now removed, so this PR sets the ground for a possible implementation where the guard condition callback is called directly from `rclcpp::GuardCondition` (instead of the originall call from RMW guard conditions).

Signed-off-by: Mauro Passerino <mpasserino@irobot.com>

irobot-ros commented on rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp:
```diff
@@ -191,17 +191,17 @@ TYPED_TEST(TestAddCallbackGroupsToExecutor, add_callback_groups_after_add_node_t
     rclcpp::CallbackGroupType::MutuallyExclusive);
   rclcpp::TimerBase::SharedPtr timer_ = node->create_wall_timer(
-    2s, timer_callback, cb_grp);
+    10ms, timer_callback, cb_grp);
```
why this change?

mauropasse replied:
This test takes 4 seconds for each executor (16 seconds when testing 4 different executors) due the period of the timer, which doesn't have any impact on the output of the test.

wjwwood replied:
I don't know why they were set to this before hand, but it might make the tests flaky. It really shouldn't, but I'm just throwing it out there, because it does happen if the test isn't constructed properly.

irobot-ros commented on rclcpp/test/rclcpp/test_node.cpp:
```diff
@@ -2556,6 +2556,7 @@ TEST_F(TestNode, get_publishers_subscriptions_info_by_topic) {
   };
   rclcpp::QoS qos = rclcpp::QoS(qos_initialization, rmw_qos_profile_default);
   auto publisher = node->create_publisher<test_msgs::msg::BasicTypes>(topic_name, qos);
+  std::this_thread::sleep_for(std::chrono::milliseconds(1));
```
this should be all synchronous, why is necessary to wait?

mauropasse replied:
I was getting random failures in this test. I didn't have it anymore when adding this sleep.

clalancette replied:
We've seen this elsewhere. The problem is that depending on the backend DDS implementation, creating a publisher and then immediately trying to get info on it may or may not work. That's because the backend may create the topic asynchronously.

You'll probably want a solution similar to #1606.

wjwwood replied:
Agreed, I'd say remove this as it is (AFAIK) unrelated to this change, and also because it isn't a scalable solution.

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -104,8 +104,8 @@ class NodeBase : public NodeBaseInterface
-  rcl_guard_condition_t *
-  get_notify_guard_condition() override;
+  rclcpp::GuardCondition::SharedPtr
+  get_notify_guard_condition() const override;
```
If you add `const` the return type should be `const` too, either `shared_ptr<const rclcpp::GuardConditioon>` or, as I suggested, `const rclcpp::GuardCondition &`.

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_base_interface.hpp:
```diff
@@ -138,12 +139,12 @@ class NodeBaseInterface
-  rcl_guard_condition_t *
-  get_notify_guard_condition() = 0;
+  rclcpp::GuardCondition::SharedPtr
+  get_notify_guard_condition() const = 0;
```
Same comments apply here as from `NodeBase`.

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_base_interface.hpp:
```diff
@@ -170,6 +171,11 @@ class NodeBaseInterface
+  RCLCPP_PUBLIC
+  virtual
+  void
+  trigger_notify_guard_condition() const = 0;
```
This needs documentation.

wjwwood commented on rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp:
```diff
@@ -504,7 +504,7 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
     std::vector<T, typename std::allocator_traits<Alloc>::template rebind_alloc<T>>;
 
-  VectorRebind<const rcl_guard_condition_t *> guard_conditions_;
+  VectorRebind<GuardCondition::SharedPtr> guard_conditions_;
```
Why does this need to be a shared pointer now? Previously it did not share ownership, is that now necessary?

wjwwood commented on rclcpp/src/rclcpp/guard_condition.cpp:
```diff
@@ -77,4 +78,15 @@ GuardCondition::exchange_in_use_by_wait_set_state(bool in_use_state)
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(
+      ret, "failed to add guard condition to wait set");
+    return false;
```
This line of code is never reached... It seems odd (maybe an accident) to mix return type and exceptions.

wjwwood reviewed (changes requested):
This pull request is definitely an improvement in a good direction.

I had concern that we're just using shared pointer everywhere, when we don't really need to, and I pointed that out in places.

Also, anywhere the API is changed in a way that would break existing code calling it in one of our `.hpp` files, we need to consider deprecating it rather than just changing it. It isn't always possible, desirable, or necessary, but we need to think about each case and justify it if we don't deprecate things.

mauropasse commented:
Last commit:
- rebased master.
- Now using `rclcpp::GuardCondition` instead of shared pointers of it.
- Added documentation to `GuardCondition::add_to_wait_set` but not sure if is enough.
- Reverted changes in tests not related to the use of `rclcpp::GuardCondition`.
- Created new api to get the node's `rclcpp::GuardCondition` (I had to slightly modify the original `NodeBase::get_notify_guard_condition`)
Let me know if I'm missing something

mauropasse commented on rclcpp/src/rclcpp/executor.cpp:
```diff
@@ -129,15 +117,9 @@ Executor::~Executor()
   // Remove and release the sigint guard condition
-  memory_strategy_->remove_guard_condition(&shutdown_guard_condition_->get_rcl_guard_condition());
+  memory_strategy_->remove_guard_condition(shutdown_guard_condition_.get());
+  memory_strategy_->remove_guard_condition(&interrupt_guard_condition_);
```
This line is new, before the `interrupt_guard_condition_` wasn't being removed from the `memory_strategy_`

irobot-ros commented:
Hi, any updates on this PR?
We think that we have addressed the review comments

wjwwood commented on rclcpp/include/rclcpp/experimental/subscription_intra_process_base.hpp:
```diff
@@ -39,8 +40,10 @@ class SubscriptionIntraProcessBase : public rclcpp::Waitable
-  : topic_name_(topic_name), qos_profile_(qos_profile)
+  SubscriptionIntraProcessBase(
+    rclcpp::Context::SharedPtr context,
+    const std::string & topic_name, rmw_qos_profile_t qos_profile)
```
```suggestion
    const std::string & topic_name,
    rmw_qos_profile_t qos_profile)
```

mauropasse replied:
Done in b363b21

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -104,9 +104,13 @@ class NodeBase : public NodeBaseInterface
 
   RCLCPP_PUBLIC
-  rcl_guard_condition_t *
+  const rcl_guard_condition_t *
```
Why is this needed? This technically breaks API.

mauropasse replied:
I removed this change in dd4ab15. For that I had to add a non-const version of `get_rcl_guard_condition` API from `GuardCondition` class

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_base_interface.hpp:
```diff
@@ -142,9 +143,20 @@ class NodeBaseInterface
   RCLCPP_PUBLIC
   virtual
-  rcl_guard_condition_t *
+  const rcl_guard_condition_t *
```
Same here, it shouldn't be necessary if you can return a non-const pointer to the rclcpp guard condition.

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_base_interface.hpp:
```diff
@@ -142,9 +143,20 @@ class NodeBaseInterface
+  const rcl_guard_condition_t *
   get_notify_guard_condition() = 0;
 
+  /// Return guard condition that should be notified when the internal node state changes.
```
```suggestion
  /// Return a guard condition that should be notified when the internal node state changes.
```

mauropasse replied:
Done in b363b21

wjwwood commented on rclcpp/include/rclcpp/executor.hpp:
```diff
@@ -548,7 +548,7 @@ class Executor
 
   typedef std::map<rclcpp::node_interfaces::NodeBaseInterface::WeakPtr,
-      const rcl_guard_condition_t *,
+      const rclcpp::GuardCondition *,
```
As long as we're breaking API, let's change this to a reference (where ever it makes sense):

```suggestion
      const rclcpp::GuardCondition &,
```

This prevents the possibility of passing a nullptr into these methods that took pointers to guard conditions in the past. It puts the responsibility onto the caller of the methods to check this.

I think this is way more idiomatic C++, and I'd prefer to do it now, as we upgrade to C++ classes from C structs, instead of doing it in two steps, breaking API each time.

mauropasse replied:
With my last commit, the API is not being broken anymore. Anyway I tried this change, but having a reference in a map seems not behave nicely, as references must be init directly and can't be copied, I was getting a bunch of strange errors.

wjwwood replied:
Ah sorry, maybe the map should hold the pointer still, but in places where we pass by pointer, we could pass by reference instead (only in function/method interfaces).

wjwwood replied:
> the API is not being broken anymore

Also, we are still breaking API since we're changing the type from `rcl_guard_condition_t` to `rclcpp::GuardCondition` in several places in the API. Which is fine, but anyway it is breaking.

mauropasse replied:
Done in ccc78c5

wjwwood commented on rclcpp/include/rclcpp/executors/static_executor_entities_collector.hpp:
```diff
@@ -65,7 +65,7 @@ class StaticExecutorEntitiesCollector final
     rcl_wait_set_t * p_wait_set,
     rclcpp::memory_strategy::MemoryStrategy::SharedPtr memory_strategy,
-    rcl_guard_condition_t * executor_guard_condition);
+    rclcpp::GuardCondition * executor_guard_condition);
```
Another example of this here, but there are several more, so I won't point them each out.

wjwwood replied:
For example, changing this one, but not the map type, would be good I think.

mauropasse replied:
Done in ccc78c5

wjwwood commented:
This needs to be rebased, due to merge conflicts, and I left some comments with requested changes, but otherwise it looks really close. Thanks for following up on it.

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -107,6 +107,10 @@ class NodeBase : public NodeBaseInterface
   get_notify_guard_condition() override;
 
+  RCLCPP_PUBLIC
+  rclcpp::GuardCondition *
```
This is another case where I think we should use a reference rather than a pointer, since it should never be nullptr, right?

mauropasse replied:
Done in ccc78c5

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -107,6 +107,10 @@ class NodeBase : public NodeBaseInterface
 
+  RCLCPP_PUBLIC
+  rclcpp::GuardCondition *
+  get_notify_rclcpp_guard_condition() override;
```
Normally, I'd agree with adding a new function with the new return type, but since we're changing (in several places) public API to take an rclcpp guard condition instead of an rcl one, maybe we should just update the existing `get_notify_guard_condition()` to return the rclcpp guard condition instead.

At the very least we should deprecate `rcl_guard_condition_t * get_notify_guard_condition()` if we don't just change the return type.

Instead, users can get the rclcpp guard condition and then get the rcl one from that, if they still need it.

mauropasse replied:
> maybe we should just update the existing get_notify_guard_condition() to return the rclcpp guard condition instead.

Done in 2b2cda5

wjwwood commented on rclcpp/include/rclcpp/guard_condition.hpp:
```diff
@@ -89,6 +95,15 @@ class GuardCondition
+   */
+  RCLCPP_PUBLIC
+  void
+  add_to_wait_set(rcl_wait_set_t * wait_set) const;
```
I'm reluctant to add new public API that uses the rcl type here. Ideally, we'd update it to work with the `rclcpp::WaitSet`, as I've said elsewhere.

Since that it out of scope, if we still want this function, I'd like to put it into a free function, in the detail namespace, so it is not a public API, and therefore we can remove it later when we roll out use of the rclcpp WaitSet.

So something like this:

```c++
// rclcpp/include/rclcpp/detail/add_guard_condition_to_rcl_wait_set.hpp
namespace rclcpp
{
namespace detail
{
RCLCPP_PUBLIC
void
add_guard_condition_to_rcl_wait_set(rcl_wait_set_t & wait_set, const rclcpp::GuardCondition & guard_condition);
}  // namespace detail
}  // namespace rclcpp

mauropasse replied:
Done in 0d6e8a9

wjwwood commented on rclcpp/include/rclcpp/memory_strategy.hpp:
```diff
@@ -64,9 +64,11 @@ class RCLCPP_PUBLIC MemoryStrategy
 
-  virtual void add_guard_condition(const rcl_guard_condition_t * guard_condition) = 0;
+  virtual void
+  add_guard_condition(const rclcpp::GuardCondition * guard_condition) = 0;
```
This is another place where we could take a reference instead of a pointer.

mauropasse replied:
Done in ccc78c5

wjwwood commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -310,3 +297,10 @@ NodeBase::resolve_topic_or_service_name(
   return output;
 }
+
+void
+NodeBase::trigger_notify_guard_condition()
+{
+  std::unique_lock<std::recursive_mutex> lock(notify_guard_condition_mutex_);
+  notify_guard_condition_.trigger();
+}
```
With the introduction of this function, maybe we don't need the `acquire_notify_guard_condition_lock` any longer and should deprecate it.

I can't find why we had it this way, it seemed intended to be used like this:

```c++
  {
    auto notify_condition_lock = node_base_->acquire_notify_guard_condition_lock();
    // some other "critical section" stuff
    rcl_ret_t ret = rcl_trigger_guard_condition(node_base_->get_notify_guard_condition());
    if (RCL_RET_OK != ret) {
      throw_from_rcl_error(ret, "failed to trigger notify guard condition");
    }
  }
```

But it doesn't appear to be used like that anywhere. At least not anymore.

And now I'm wondering if the lock is needed at all. I think triggering a guard condition should be thread-safe...

mauropasse replied:
> With the introduction of this function, maybe we don't need the acquire_notify_guard_condition_lock any longer and should deprecate it.

> I'm wondering if the lock is needed at all. I think triggering a guard condition should be thread-safe

I also don't think the lock is needed, as triggering the guard condition is thread-safe.
Done in 82629b7

wjwwood commented on rclcpp/test/rclcpp/node_interfaces/test_node_services.cpp:
```diff
@@ -103,7 +103,7 @@ TEST_F(TestNodeService, add_service_rcl_trigger_guard_condition_error)
     "lib:rclcpp", rcl_trigger_guard_condition, RCL_RET_ERROR);
   RCLCPP_EXPECT_THROW_EQ(
     node_services->add_service(service, callback_group),
-    std::runtime_error("Failed to notify wait set on service creation: error not set"));
```
This seems like a step backwards... this is, perhaps, another reason to not introduce the `trigger_notify_guard_condition()` method?

Or it should have an error message argument to augment this error message. Or maybe all the callers should catch and rethrow. Either way, it would be better if these errors didn't change in this way.

mauropasse replied:
I'm not sure of the correct or best way to implement the augment of the error message. Should be passed as argument when calling `guard_condition->trigger("some error message")` ?

Regarding the introduction of `trigger_notify_guard_condition`, I can always remove it, so first we have to get the node's notify guard condition and then trigger it.

wjwwood replied:
I would say we should catch the throw from trigger and rethrow with more information.

mauropasse replied:
Done in 765257c

wjwwood commented on rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp:
```diff
@@ -61,24 +62,24 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
-        break;
-      }
-    }
+    add_guard_condition(guard_condition.get());
```
Since this method is not retaining the ownership of the shared pointer (or otherwise using it), can we just get rid of these signatures that take the shared pointer and leave just the ones that take a reference?

mauropasse replied:
Done in 7a4b2a1

wjwwood commented on rclcpp/src/rclcpp/executor.cpp:
```diff
@@ -699,9 +668,9 @@ Executor::wait_for_work(std::chrono::nanoseconds timeout)
+            const auto & guard_condition = node_guard_pair->second;
             weak_nodes_to_guard_conditions_.erase(weak_node_ptr);
-            memory_strategy_->remove_guard_condition(guard_condition);
+            memory_strategy_->remove_guard_condition(*guard_condition);
```
Is it possible to have this be nullptr unexpectedly? Should we check it?

mauropasse replied:
Here we're in the situation in which a node could be expired: `weak_node_ptr.expired()`
If that the case the node's guard condition has been destructed, so the dereference `*guard_condition` is wrong since is now a dangling pointer. It could be `nullptr` or not, either way is dangerous.
The memory strategy should hold weak pointers of the guard condition, or have a map `{WeakNode, GuardCondition}` as the executor does, or share the same map. 
So the node's GuardCondition should be created as SharedPtr, and stored as WeakPtr in executor/memory_strategy. 
What do you think?

irobot-ros replied:
What about storing only WeakNodes? I assume that regardless of the approach, before using the guard condition we will still have to check that the node has not gone out of scope, so at that point it may be worth directly calling `get_notify_guard_condition` every time

mauropasse replied:
Commit 9da50f1 fixes this problem (accessing an out-of-scope `GuardCondition` pointer) by restoring `remove_guard_condition` original way of taking directly a pointer to the guard condition to remove.
This way we also avoid to take the address of a `GuardCondition` inside the API, which might have gone out of scope.
I have rebased this PR with master. I think all the comments have been addressed now.

wjwwood commented on rclcpp/src/rclcpp/subscription_intra_process_base.cpp:
```diff
@@ -21,8 +22,9 @@ SubscriptionIntraProcessBase::add_to_wait_set(rcl_wait_set_t * wait_set)
-  return RCL_RET_OK == ret;
+  detail::add_guard_condition_to_rcl_wait_set(*wait_set, gc_);
+
+  return true;
```
Should we just change this to not return anything?

mauropasse replied:
Done in 34f8ed0

wjwwood commented on rclcpp/test/rclcpp/executors/test_executors.cpp:
```diff
@@ -352,40 +354,18 @@ TYPED_TEST(TestExecutors, testSpinUntilFutureCompleteWithTimeout) {
-    rcl_ret_t ret = rcl_wait_set_add_guard_condition(wait_set, &gc_, NULL);
-    return RCL_RET_OK == ret;
+    rclcpp::detail::add_guard_condition_to_rcl_wait_set(*wait_set, gc_);
+    return true;
```
Same here, should we just not return anything?

mauropasse replied:
Done in 34f8ed0

wjwwood reviewed (commented):
A few more comments and suggestions, but this is really close!

Thanks for your patience in iterating on this.

mauropasse commented on rclcpp/include/rclcpp/guard_condition.hpp:
```diff
@@ -59,6 +61,10 @@ class GuardCondition
+  rcl_guard_condition_t &
+  get_rcl_guard_condition();
+
   RCLCPP_PUBLIC
```
Done in 7a4b2a1

wjwwood commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -265,20 +242,14 @@ NodeBase::get_associated_with_executor_atomic()
   std::lock_guard<std::recursive_mutex> notify_condition_lock(notify_guard_condition_mutex_);
   if (!notify_guard_condition_is_valid_) {
-    return nullptr;
+    throw std::runtime_error("Trying to get invalid notify guard condition");
```
```suggestion
    throw std::runtime_error("failed to get notify guard condition because it is invalid");
```

mauropasse replied:
Done in f99baec

wjwwood commented on rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:
```diff
@@ -348,12 +348,12 @@ NodeGraph::notify_graph_change()
+    node_gc.trigger();
+  } catch (const rclcpp::exceptions::RCLError & ex) {
+    throw std::runtime_error(
+            std::string("Failed to notify wait set on graph change: ") + ex.what());
```
```suggestion
            std::string("failed to notify wait set on graph change: ") + ex.what());
```

mauropasse replied:
Done in f99baec

wjwwood commented on rclcpp/src/rclcpp/node_interfaces/node_services.cpp:
```diff
@@ -41,14 +41,12 @@ NodeServices::add_service(
+    node_gc.trigger();
+  } catch (const rclcpp::exceptions::RCLError & ex) {
+    throw std::runtime_error(
+            std::string("Failed to notify wait set on service creation: ") + ex.what());
```
I know it was like this before, but the predominate style elsewhere is lower case:

```suggestion
            std::string("failed to notify wait set on service creation: ") + ex.what());
```

mauropasse replied:
Done in f99baec

wjwwood commented on rclcpp/src/rclcpp/node_interfaces/node_services.cpp:
```diff
@@ -68,14 +66,12 @@ NodeServices::add_client(
+    node_gc.trigger();
+  } catch (const rclcpp::exceptions::RCLError & ex) {
+    throw std::runtime_error(
+            std::string("Failed to notify wait set on client creation: ") + ex.what());
```
```suggestion
            std::string("failed to notify wait set on client creation: ") + ex.what());
```

mauropasse replied:
Done in f99baec

wjwwood commented on rclcpp/src/rclcpp/node_interfaces/node_timers.cpp:
```diff
@@ -41,11 +41,15 @@ NodeTimers::add_timer(
     throw std::runtime_error(
-            std::string("Failed to notify wait set on timer creation: ") +
-            rmw_get_error_string().str);
+            std::string("Failed to notify wait set on timer creation: ") + ex.what());
```
```suggestion
            std::string("failed to notify wait set on timer creation: ") + ex.what());
```

mauropasse replied:
Done in f99baec

wjwwood commented on rclcpp/src/rclcpp/node_interfaces/node_topics.cpp:
```diff
@@ -60,13 +60,12 @@ NodeTopics::add_publisher(
+    node_gc.trigger();
+  } catch (const rclcpp::exceptions::RCLError & ex) {
+    throw std::runtime_error(
+            std::string("Failed to notify wait set on publisher creation: ") + ex.what());
```
```suggestion
            std::string("failed to notify wait set on publisher creation: ") + ex.what());
```

mauropasse replied:
Done in f99baec

wjwwood commented on rclcpp/src/rclcpp/node_interfaces/node_topics.cpp:
```diff
@@ -108,13 +107,12 @@ NodeTopics::add_subscription(
+    node_gc.trigger();
+  } catch (const rclcpp::exceptions::RCLError & ex) {
+    throw std::runtime_error(
+            std::string("Failed to notify wait set on subscription creation: ") + ex.what());
```
```suggestion
            std::string("failed to notify wait set on subscription creation: ") + ex.what());
```

mauropasse replied:
Done in f99baec

wjwwood commented on rclcpp/src/rclcpp/node_interfaces/node_waitables.cpp:
```diff
@@ -41,14 +41,12 @@ NodeWaitables::add_waitable(
+    node_gc.trigger();
+  } catch (const rclcpp::exceptions::RCLError & ex) {
+    throw std::runtime_error(
+            std::string("Failed to notify wait set on waitable creation: ") + ex.what());
```
```suggestion
            std::string("failed to notify wait set on waitable creation: ") + ex.what());
```

mauropasse replied:
Done in f99baec

wjwwood commented on rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp:
```diff
@@ -326,9 +326,9 @@ TEST_F(TestNodeGraph, notify_graph_change_rcl_error)
+  RCLCPP_EXPECT_THROW_EQ(
     node()->get_node_graph_interface()->notify_graph_change(),
-    rclcpp::exceptions::RCLError);
+    std::runtime_error("Failed to notify wait set on graph change: error not set"));
```
This seems odd... why is it "error not set"?

wjwwood replied:
I guess it was like that before... but I'm not sure.

wjwwood reviewed (commented):
Sorry for the long delay. Had a few more comments/questions. But the changes that you made in response to my last review looked good.

mauropasse commented on rclcpp/include/rclcpp/detail/add_guard_condition_to_rcl_wait_set.hpp:
```diff
@@ -0,0 +1,40 @@
+
+/// Adds the guard condition to a waitset
+/**
+ * This function is thread-safe.
```
Yes you are right, there could be data races on the `wait_set`. I've removed the statement.

mauropasse commented:
@wjwwood I think I fixed the test, this PR is needed by `test_rclcpp`: https://github.com/ros2/system_tests/pull/488

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -114,13 +114,9 @@ class NodeBase : public NodeBaseInterface
 
   RCLCPP_PUBLIC
-  rcl_guard_condition_t *
+  rclcpp::GuardCondition &
   get_notify_guard_condition() override;
```
This API also should be mentioned in the notice we put in the Humble release page. It is even less likely to be used outside of the core, but still it should be mentioned.

mauropasse replied:
Done in https://github.com/ros2/ros2_documentation/pull/2086

wjwwood reviewed (approved):
lgtm, just need to sort out what to do/document for the API  breaks.

I'll run CI again.

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15546)](http://ci.ros2.org/job/ci_linux/15546/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10230)](http://ci.ros2.org/job/ci_linux-aarch64/10230/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13213)](http://ci.ros2.org/job/ci_osx/13213/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15757)](http://ci.ros2.org/job/ci_windows/15757/)

mauropasse commented on rclcpp/include/rclcpp/waitable.hpp:
```diff
@@ -103,12 +103,11 @@ class Waitable
   virtual
-  bool
+  void
   add_to_wait_set(rcl_wait_set_t * wait_set) = 0;
```
Done in https://github.com/ros2/ros2_documentation/pull/2086

mauropasse commented:
Just in case the `rclcpp_test` errors I see in the CI about `test_two_service_calls__rmw_*`, are not related with changes from this PR. I can reproduce them locally using the master branch.
Anything missing from this PR @wjwwood ?

wjwwood commented:
@mergifyio rebase
